### ISSUE TITLE
torch.cuda.amp issue with torchvision densenet121 when memory_efficient=True

### ISSUE THREAD

**Issue by andwurl** [pytorch/vision#3122]
## 🐛 Bug

I have been using a torchvision densenet121 with PT1.4+ APEX/AMP for mixed precision training. Today, I tried to migrate my code to PT1.7 and to native mixed precision (amp). 

In short, when using a densenet121 with memory_efficient=True, I get an amp related error message: 'Input type (torch.cuda.HalfTensor) and weight type (torch.cuda.FloatTensor) should be the same'

When using the densenet121 with memory_efficient=False, everything works fine and in mixed precision mode.

## To Reproduce
import torchvision.models as models
(A) model=models.densenet121(pretrained=None, memory_efficient=True)
(B) model=models.densenet121(pretrained=None, memory_efficient=False)

for epoch in epochs:
    for input, target in data:
        optimizer.zero_grad()

        with autocast():
            output = model(input)
            loss = loss_fn(output, target)

        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()

How could the densenet(memory_efficient=True) be modified so that it becomes compatible with mixed precision training?

## Environment
PyTorch version: 1.7.0+cu101
Is debug build: True
CUDA used to build PyTorch: 10.1
OS: SUSE Linux Enterprise Server 12 SP5 (x86_64)
GCC version: (SUSE Linux) 4.8.5
Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
Nvidia driver version: 418.39
[pip] torch==1.7.0+cu101
[pip] torchaudio==0.7.0
[pip] torchtuples==0.2.0
[pip] torchvision==0.8.0

**Comment by vfdev-5** [comment]
Related to pytorch issue https://github.com/pytorch/pytorch/issues/47183 , here is some (unrecommended) workaround: https://github.com/pytorch/pytorch/issues/37730#issuecomment-623296247

**Comment by andwurl** [comment]
Thanks a lot!
Running scaler.scale(loss).backward() inside the autocast block helps, as well as the changes suggested in #37730 to checkpoint.py. However, I have observed in both cases that memory consumption was higher compared to PT1.4+APEX/AMP.
